---
title: "R Notebook"
output: html_notebook
---


```{r}
#library(phyloseq)
library(dplyr)
library(here)
library(here) #to construct a path
library(readr)
library(readxl)
library(tidyr)
library(DECIPHER)
library(ggplot2)
library(vegan)
library(forcats)
library(eulerr)
```


```{r}
# ASV = Amplicon sequence variant
seqtab.nochim <- read.table(here::here("Data", "COI_ASV_table_99.text"))
# Sequence table from DADA2 (without chimeras) 
# Contains the total amount of reads for each sequence within each sample

seqtab.nochim
# Displays theASV table


# This code is for the old ASV table that didnt have the column names renamed:
# colnames(seqtab.nochim) <- paste("seq", formatC(1:ncol(seqtab.nochim), flag = "0", width = nchar(ncol(seqtab.nochim))), sep = "_")
# Changing the column names. Giving a sequence number to the actual sequence (ACTGCTA...)

# The old ASV table also took a long time to read in so to avoid the long waiting time it was saved as an RDS file and then the RDS file was loaded in, which was much quicker:
# saveRDS(seqtab.nochim, here("Data", "seqtab.nochim.RDS"))
# Saves the seqtab.nochim in another format that is faster to open

# Just load this after you have saved it once
#seqtab.nochim <- readRDS(here("Data", "seqtab.nochim.RDS"))
```

Tax table - CHECK!
```{r}
# Load in the Tax table from DADA2:
# Reference database: Nordic countries:
COI_blast_N <- read_tsv(here::here("Data", "COI_blast_embl.txt"), col_names = FALSE)

colnames(COI_blast_N) <- c("ASV","matchID", "pident","qstart", "qend","bitscore","length","nident")

ranks <- c( "Phylum", "Class", "Order", "Family", "Subfamily", "Genus", "Species")

Tax_table_95 <- 
  COI_blast_N %>%  
  separate(matchID, ranks, sep = ";") %>% 
  mutate_if(is.character, function(x){case_when(x == "NA" ~ NA_character_,
                                                TRUE ~ x)}) %>% 
  mutate(n_ranks = rowSums(!is.na(.[,ranks]))) %>% 
  group_by(ASV) %>% 
  dplyr::filter(bitscore > max(bitscore)*0.97) %>% 
  dplyr::filter(n_ranks >= max(n_ranks)-2) %>% 
  dplyr::filter(bitscore == max(bitscore)) %>% 
 #  dplyr::filter(pident > 90) %>% 
  dplyr::slice(1) %>% 
 #  mutate(Species = case_when(pident < 97 ~ NA_character_,
 #                             grepl("sp._", Species) ~ NA_character_,
 #                             TRUE~ Species)) %>% 
 #  mutate(Genus = case_when(pident < 95 ~ NA_character_,
 #                             TRUE~ Genus)) %>% 
  select(-Subfamily) %>% 
  mutate(Phylum = case_when(Class == "Oomycota" ~ "Heterokontophyta",
                            Class == "Xanthophyceae" ~ "Ochrophyta",
                            Class == "Phaeophyceae" ~ "Ochrophyta",
                            Class == "Chrysophyceae" ~ "Ochrophyta",
                            TRUE ~ Phylum))

kingdom_pred <- read_tsv(here("Data", "Kingdom_prediction_COI99.txt"))

Tax_animalia <- 
Tax_table_95 %>% 
  left_join(kingdom_pred) %>% 
  filter(Kingdom == "animalia") %>% 
  filter(Phylum %in% c("Arthropoda", "Chordata")) 
  
Tax_animalia %>% 
  group_by(Kingdom, Phylum, Class, Order) %>% 
  summarise(n = n()) %>% 
  filter(Phylum == "Arthropoda" & n > 2 | Phylum == "Chordata") %>% 
  filter(Order != "Primates") %>% 
  group_by(Class) %>% 
  mutate(nc = n()) %>% 
  arrange(desc(n)) %>% 
  ungroup() %>% 
  mutate(Class = factor(Class, levels = unique(.$Class))) %>% 
  arrange(n) %>% 
  mutate(Order = factor(Order, levels = .$Order)) %>% 
  filter(!is.na(Order)) %>% 
  ggplot(aes(x = Order, y = n, fill = Class))+
  geom_bar(stat = "identity")+
  geom_label(aes(y= 0.5*n, label = n), label.padding = unit(0.08, "lines"), 
             fill = "white", alpha = 0.5, colour = NA)+
  geom_text(aes(y = 0.5*n, label = n), size = 3)+
  annotate(geom = "text", y = 16, x = 14, 
           label = "total (Order) :\nArthropods : 201(26)\nChordata : 16(11)",
           hjust = 0, size = 6)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))+
  scale_fill_brewer(palette = "Paired")+
  coord_flip()+
  theme_bw()+
  guides(fill =  guide_legend(ncol = 2))+
  theme(legend.position = c(0.7, 0.2),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())+
  labs(y = "number of detected species")


ggsave(here("Figures", "Order.pdf"), width = 6, height = 4)

Tax_table_95 %>% 
  ungroup() %>% 
  left_join(kingdom_pred) %>% 
  filter(Kingdom == "animalia") %>% 
  filter(Phylum %in% c("Arthropoda", "Chordata")) %>% 
  filter(Order != "Primates") %>% 
  select(Phylum, Order) %>% 
  distinct() %>% 
  group_by(Phylum) %>% 
  summarise(n = n())

Tax_table_95 %>% 
  left_join(kingdom_pred) %>% 
  select(Kingdom, Phylum) %>% 
  filter(Kingdom == "fungi", Phylum %in% c("Ascomycota", "Basidiomycota")) %>% 
  group_by(Phylum) %>% summarise(n = n())
```

```{r}
COIseq <- readDNAStringSet("Data/COI_ASVs.fa") 

COIseq["seq_0038"] %>% as.character()
```


Metadata, sampleInfo: - CHECK!
```{r}
sampleInfo <- read_tsv(here::here("Data", "Sequencing_data", "P14711", "00-Reports", "Y.Clough_19_01_sample_info.txt"))

# Read in file with sample info, from DADA2/NGI(?)

sampleInfo <- 
  sampleInfo %>% 
  mutate(Barcode = gsub("(.+)_.+", "\\1", `User ID`),
         ID = gsub(".+_(.+)", "\\1", `User ID`),
         Station = gsub("([A-Z]+)[0-9]{0,2}", "\\1", `ID`))

# Adds 3 new columns to the dataset "Barcode" (COI or 16S), "ID" (SB1, SB2...) and "Station" (SB, SF, SI, IBA...)
# mutate function creates new variables from a dataset
# gsub function specifys what you take from the userID column and put into the new column

sampleInfo <- filter(sampleInfo, Barcode == "COI")
# Only shows data for COI

sampleInfo$Station %>% unique()
# Tells the different unique values that are available in column "Station", which are 7: SB, SF, SI, IBA, ECO, TEST and CLEAN

sampleInfo_noIBA <- filter(sampleInfo, Station != "IBA")

sampleInfo
# Displays sampleInfo

```

ASV table
```{r}
seqtab.nochim <- readRDS(here::here("Data", "seqtab.nochim.RDS"))
seqtab.nochim99 <- read.table(here::here("Data", "COI_ASV_table_99.text"))
```

```{r}
#seqtab.nochim99[seqtab.nochim99 > 0] <- 1
#seqtab.nochim[seqtab.nochim > 0] <- 1

lapply(rownames(seqtab.nochim), function(x){
  
  S199 <- colnames(seqtab.nochim99)[seqtab.nochim99[x,] > 0]
  S1 <- colnames(seqtab.nochim)[seqtab.nochim[x,] > 0]
  seqdiff <- S199[which(! S199 %in% S1)]
  Tax_table_95 %>% filter(ASV %in% seqdiff) %>% pull(ASV)
  
}
       )

x <- rownames(seqtab.nochim)[23]

S199[which(! S199 %in% S1)]

seqtab.nochim99[1,S199[which(! S199 %in% S1)]]

COI_blast_N %>% filter(ASV %in% seqdiff)

seq006 <- filter(Cluster_ID, match %in% seqdiff[3]) %>% pull(seq)

unique(COI_blast_N$ASV)
filter(COI_blast_N, ASV %in% seqdiff[3])

seqtab_fasta <- readDNAStringSet(here("Data", "COI_ASVs.fa"))

seqtab_fasta[c("seq_0058", "seq_0419", "seq_1228", "seq_2745", "seq_3194")] %>% as.character()

COI_blast_N %>% filter(ASV %in% seq006) #%>% 
  mutate_at(vars(matches(ranks)), function(x) {
    case_when(x == "NA" ~ NA_character_,
              TRUE ~ x)
  }) %>%  
  filter(pident > 95) %>% 
  # at each taxonomic rank take the absolut majority rank (see majority_tax() )
  mutate_at(vars(matches(ranks)), majority_tax)
  
  Tax_table_95 %>% filter(ASV %in% c("seq_0058", "seq_0006", "seq_0018"))
```


Tax table
```{r}
COI_blast_N <- read_tsv(here("Data", "COI_Bold_Nordic_out.txt"), col_names = FALSE)
COI_blast_N <- read_tsv(here("Data", "COI_BOLD_Blast_out.txt"), col_names = FALSE)

COI_blast_N <- COI_blast_N[,-14]

colnames(COI_blast_N) <- c("ASV","matchID","pident","length",
                         "mismatch","gapopen","qstart","qend",
                         "sstart","send","evalue","bitscore",
                         "qlen")

#subset for only ASVs in clustered ASV table
COI_blast_N <- COI_blast_N %>% filter(ASV %in% colnames(seqtab.nochim99))

ranks <- c( "Phylum", "Class", "Order", "Family", "Genus", "Species")

COI_blast_N <- 
COI_blast_N %>% 
  separate(matchID, ranks, sep = ";")


majority_tax <- function(x) {
  
  if(sum(!is.na(x)) == 0) return(NA)
  
  freq <- table(x)
  max_freq <- freq[freq == max(freq)]
  
  if(length(max_freq) > 1) return(NA)
  if(max_freq < sum(freq)/2) return(NA)
  return(names(max_freq))
}

Tax_table_95 <- 
  COI_blast_N %>%
  dplyr::filter(pident > 90) %>% 
  #replace "NA" with NA
  mutate_at(vars(matches(ranks)), function(x) {
    case_when(x == "NA" ~ NA_character_,
              TRUE ~ x)
  }) %>%  
  group_by(ASV) %>% 
  #filter best hits (highest pident + hits with < 2% lower pident)
  filter(pident > max(pident -2)) %>% 
  
  # at each taxonomic rank take the absolut majority rank (see majority_tax() )
  mutate_at(vars(matches(ranks)), majority_tax) %>% 
  select(1:7) %>% 
  distinct()

Tax_table_95 %>% 
  group_by(Class) %>% 
  summarise(n = n()) %>% 
  arrange(desc(n)) %>% View()


Tax_table_95 %>% 
  filter(Phylum == "Cnidaria") 

COI_ASVs <- readDNAStringSet(here("Data", "COI_clustered_99.fasta"))

as.character(COI_ASVs[c("seq_0019")])

filter(Tax_table_95, ASV == "seq_0019")

```

metadata
```{r}

sampleInfo <- read_tsv(here::here("Data", "Sequencing_data", "P14711", "00-Reports", "Y.Clough_19_01_sample_info.txt"))

sampleInfo <- 
  sampleInfo %>% 
  mutate(Barcode = gsub("(.+)_.+", "\\1", `User ID`),
         ID = gsub(".+_(.+)", "\\1", `User ID`),
         Station = gsub("([A-Z]+)[0-9]{0,2}", "\\1", `ID`)) %>% 
  mutate(Exp = case_when(grepl("S", Station) ~ "Bees_and_Butterflies",
                         Station %in% c("ECO", "TEST") ~ "Moths",
                         TRUE ~ Station))

sampleInfo <- filter(sampleInfo, Barcode == "COI")

sampleInfo$Station %>% unique()

clean_seq <- 
colnames(seqtab.nochim)[which(seqtab.nochim[sampleInfo[grepl("ECO", sampleInfo$Station),]$`NGI ID`,] > 0)]

filter(Tax_table_95, ASV %in% clean_seq) 

  sampleInfo$Station %>% unique()
  
  COI_blast %>% filter(! ASV %in%  COI_blast_N$ASV) 
```

#Data descritpion

```{r}
#prct of reads assigned at all during BLAST
{sum(seqtab.nochim[,  colnames(seqtab.nochim) %in% COI_blast_N$ASV])/
  sum(seqtab.nochim)*100} 

#prct ASVs assigned at all during BLAST 
length(unique(COI_blast_N$ASV)) / ncol(seqtab.nochim) * 100

```
only about 30% of all reads and 25% of all ASVs could had a match of >= 80% in our BLAST database 


```{r}
{sum(seqtab.nochim[,  colnames(seqtab.nochim) %in% Tax_table_95$ASV])/
  sum(seqtab.nochim[,  colnames(seqtab.nochim) %in% COI_blast_N$ASV])*100}

length(Tax_table_95$ASV) / length(unique(COI_blast_N$ASV))*100
```
Of the ones with a hit, about 60% of the reads and 30% of the ASVs have a match with 90% identity or more


How many Arthropods / sample

```{r}


Phylum_abund <- 
Tax_table_95 %>% 
  split(.$Phylum) %>% 
  lapply(., function(x) rowSums(seqtab.nochim[, x$ASV])) %>% 
  bind_rows() %>% 
  mutate(ID = sampleInfo[match(rownames(seqtab.nochim), 
                               sampleInfo$`NGI ID`),]$ID) %>%
  gather(Phylum, abund, -ID) 

P_ab <- 
Phylum_abund %>% 
  group_by(ID) %>% 
  left_join(sampleInfo) %>% 
  filter(Exp != "CLEAN") %>% 
  mutate(abund = signif((abund / sum(abund))*100,2)) %>%
  ggplot(aes(x = Phylum, y = abund, colour = Exp))+
geom_point(position = position_jitterdodge(jitter.width = 0.1,dodge.width = 0.5))+
  facet_wrap(~Exp)+
 # stat_summary(fun.y = "mean", colour = "black", geom = "point", size = 2)+
  theme_bw()+
  labs(title = "",
       y = "% of reads",
       x= "")+
  scale_color_brewer(palette = "Set1")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  theme(legend.position = "none")

P_ab


  Tax_table_95 %>% filter(Phylum == "Heterokontophyta")
  
```


filtering out the Insects & building a phyloseq object 

We also collapse the species with same taxonomic assignment down to Species level. 
we check that the distance matrix seems reasonable for the same species which it does (<3%)

```{r}
Tax_table_95 <- filter(Tax_table_95, Class == "Arthropoda")

Tax_table_95

mult_sp <- 
Tax_table_95 %>% 
  group_by_at(ranks) %>% 
  summarise(n =n()) %>% 
  arrange(desc(n)) %>% 
  filter(n > 1) 

distMat <- 
filter(Tax_table_95, Genus %in% mult_sp$Genus) %>% 
  split(.$Genus) %>% 
  lapply(., function(x) {
    COI_fasta[x$ASV] %>% 
      AlignSeqs(verbose = F) %>% 
      DistanceMatrix(verbose = F)*100 
    })

Dist_within <- lapply(distMat, rowMeans)

#collapse seqtab.nochim

seqtab.nochim.c <- seqtab.nochim

for( i in seq_along(Dist_within)) {
  x <- Dist_within[[i]]
    seqtab.nochim.c[,names(x)[1]] <- rowSums(seqtab.nochim.c[,names(x)])
    seqtab.nochim.c <- seqtab.nochim.c[,!colnames(seqtab.nochim.c) %in% names(x)[-1]]
  }

Tax_table_95M <- as.matrix(Tax_table_95[,-1])
rownames(Tax_table_95M) <- Tax_table_95$ASV

sampleInfoM <- as.matrix(sampleInfo)
rownames(sampleInfo) <- sampleInfo$`NGI ID`

ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE),
               tax_table(Tax_table_95M),
               sample_data(sampleInfo))

ps_pa <- ps
otu_table(ps_pa)@.Data <- decostand(otu_table(ps_pa)@.Data, "pa")

ps_barplot <- 
tax_glom(ps_pa, taxrank="Order") %>% 
 # prune_samples(sample_data(ps)$Station != "SI", .) %>% 
plot_bar(., fill="Order", x = "ID")

bp <- 
ps_barplot$data %>% 
  group_by(Order) %>%
  summarise(Abundance = sum(Abundance)) %>% 
  mutate(Order = fct_reorder(Order, Abundance)) %>%
  mutate(fct_level = as.numeric(Order)) %>% 
  arrange(desc(fct_level)) %>% 
  ggplot(aes(x = Order, y = Abundance, fill = Order))+
  geom_bar(stat = "identity", position="dodge")+
  geom_text(aes(label = paste(Order," (", Abundance,")", sep = ""),
                  y = Abundance+5*0.8), 
           fill = "NA", hjust = 0)+
  #facet_wrap(~Exp, scales = "free")+
  coord_flip()+
  theme_classic()+
  theme(legend.position = "none",
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        axis.line.y = element_blank(),
        axis.line.x = element_blank())+
  labs(title = "total detected & identified species")+
  scale_y_continuous(limits = c(0,90))


bp  +scale_fill_viridis_d()
```


# comparision with field data

## moths survey
```{r}
Moths <- read_xlsx(here::here("Data", "Moths_survey_2019_LU.xlsx"), 
                   skip = 2, guess_max = Inf)
Moths <- 
Moths %>% 
  select("Artnamn", "Vetenskapligt namn", 
         "Antal", "Metod", "Startdatum", "Slutdatum") %>% 
  filter(Metod == "lampa") %>% 
  mutate(Startdatum = ymd(Startdatum)) %>% 
  filter(Startdatum > ymd("2019-08-24") & 
           Startdatum <= ymd("2019-09-09")) %>% 
  mutate(Antal = as.numeric(Antal)) %>% 
  rename(SciName = `Vetenskapligt namn`)

Moths_sp <- Moths$SciName %>% unique

```


```{r}
Moths_samples <- sampleInfo[sampleInfo$Exp == "Moths",]$`NGI ID`
Moths_seqtab <- seqtab.nochim.c[Moths_samples, ]
Moths_seqtab <- Moths_seqtab[,colSums(Moths_seqtab) > 0]

Tax_table_95_Moths <- 
Tax_table_95 %>% 
  filter(ASV %in% colnames(Moths_seqtab))

Tax_table_95_Moths <- 
Tax_table_95_Moths %>%
 # filter(Order == "Lepidoptera") %>% 
  mutate(sp_in_list = paste(Genus, Species) %in% Moths_sp) %>% 
  mutate(genus_in_list = Genus %in% gsub("(\\w+)\\s.+","\\1", Moths_sp))

  
A = na.omit(Tax_table_95_Moths$Genus)
B = gsub("(\\w+)\\s.+","\\1", Moths_sp)
AB  = dplyr::intersect(A, B)

v <- euler(c(A=length(A), B=length(B), "A&B"=length(AB)))

plot(v, fills = alpha(c("#2b8cbe", "#ffeda0"),0.8),
     labels = c("eDNA", "light trapping"),
     quantities = TRUE)

```

## Butterfly survey (F2F)

```{r}
load(here::here("Data", "Insect_Plant.Rdata"))

```


```{r}

BB_samples <- sampleInfo[sampleInfo$Exp == "Bees_and_Butterflies",]$`NGI ID`
BB_seqtab <- seqtab.nochim.c[BB_samples, ]
BB_seqtab <- BB_seqtab[,colSums(BB_seqtab) > 0]

Tax_table_95_BB <- 
Tax_table_95 %>% 
  filter(ASV %in% colnames(BB_seqtab))

Tax_table_95_BB <- 
Tax_table_95_BB %>%
  filter(Order == "Lepidoptera") %>% 
  mutate(sp_in_list = paste(Genus, Species) %in% unique(Insects$Species)) %>% 
  mutate(genus_in_list = Genus %in% gsub("(\\w+)\\s.+","\\1", unique(Insects$Species)))

A = na.omit(Tax_table_95_BB$Genus)
B = gsub("(\\w+)\\s.+","\\1", unique(Insects$Species))
AB  = dplyr::intersect(A, B)

v <- euler(c(A=length(A), B=length(B), "A&B"=length(AB)))

plot(v, fills = alpha(c("#2b8cbe", "#ffeda0"),0.8),
     labels = c("eDNA", "transect counts"),
     quantities = TRUE)

```

```{r}

BB_seqtab[,colnames(BB_seqtab) %in% Tax_table_95_BB$ASV]


COI_fasta["seq_0038"] %>% as.character()
```

most prevalent species

```{r}
Tax_table_95C <- 
  Tax_table_95 %>% 
  group_by_at(ranks) %>% 
  summarise(n =n())
  
Seq_prev <- seqtab.nochim.c[,colnames(seqtab.nochim.c) %in% Tax_table_95$ASV]
Seq_prev[Seq_prev>0] <- 1

top10 <- sort(colSums(Seq_prev), decreasing = T)[1:10]

Tax_table_95[match(names(top10), Tax_table_95$ASV),] %>% 
  ungroup %>% 
  mutate(prevalence = top10) %>% 
  select(ASV, Genus, Species)
  pull(ASV) %>% 
  COI_fasta[.] %>% 
  as.character()
  
  COI_blast_N %>% 
    filter(pident > 95) %>% 
    filter(Phylum == "Arthropoda") %>% 
    group_by_at(ranks) %>% 
    slice(1) %>% 
    View()
    
```



```{r}
Tax_table_95
sampleInfo
seqtab.nochim

ECO_samples <- filter(sampleInfo, Station == "ECO") %>% pull(`NGI ID`)

seqtab.nochimECO <- seqtab.nochim[ECO_samples, ]

seq_ECO <- colnames(seqtab.nochim)[colSums(seqtab.nochimECO) > 0]

Tax_table_95 %>% 
  filter(Class == "Insecta") %>% 
  filter(ASV %in% seq_ECO)

subset_samples(ps, Station=="ECO")

```


```{r}

seq_incorrect <- 
tax_COI_clean %>% filter(species %in% c("Mycalesis anynana","Austrazenia pura", "Clytie illunaris")) %>% pull(seq)

prob_COI %>% filter(seq %in% c("seq_0550", "seq_0720", "seq_3659"))

COI_99[seq_incorrect] %>% 
  as.character()
                                                                                                              
```

