---
title: "Data_analysis_final"
output: html_notebook
---

```{r}
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(here)
library(ggrepel)
library(ggpubr)
library(phyloseq)
library(DESeq2)
library(readxl)
library(lubridate)
library(eulerr)
```


# download data from Figshare

all necessary datafiles to run this script are deposited on Figshare and can be downloaded here: https://doi.org/10.6084/m9.figshare.16437855


#import data

```{r}
# ASV tables clustered at 99% ID

ASV_COI <- read.table(here::here("Data", "COI_ASV_table_99.text"))
ASV_16S <- read.table(here::here("Data", "16S_ASV_table_99.text"))

# taxonomic assignment with SINTAX, and probability tables

tax_COI <- read_tsv(here::here("Data", "COI_99_Sintax_tax.txt"))
prob_COI <- read_tsv(here::here("Data", "COI_99_Sintax_prob.txt"))
colnames(prob_COI) <- tolower(colnames(prob_COI))

tax_16S <- read_tsv(here::here("Data", "16S_99_Sintax_tax.txt"))
prob_16S <- read_tsv(here::here("Data", "16S_99_Sintax_prob.txt"))
colnames(prob_16S) <- tolower(colnames(prob_16S))

# taxonomic assignment with boldigger for COI 
tax_COI_bd <- read_tsv(here("Data", "boldiger", "COI_boldiger_tax.txt"))

# sample meta-data
sampleInfo <- read_tsv(here::here("Data", "Air_DNA_meta.txt")) 

#Transect-walk Data
Insects <- read_tsv(here("Data", "Insects.txt"))

#Moth trapping data
Moths <- read_tsv(here("Data", "Moths.txt"))
```

# clean taxonomy

We remove all taxonomic assignment with a bootstrap probability of below 60%

```{r}

tax_COI_60 <- tax_COI
tax_COI_60[prob_COI < 0.6] <- NA_character_ 

tax_16S_60 <- tax_16S
tax_16S_60[,-2][prob_16S < 0.6] <- NA_character_

```


# remove unassigned sequences

```{r}

## COI
unassigned_coi <- 
  filter(tax_COI_60, is.na(phylum)) %>% 
  pull(seq)

#proportion of sequences
round((length(unassigned_coi) / nrow(tax_COI_60))*100, 1)

#proportion of reads
round((sum(ASV_COI[,unassigned_coi]) / sum(ASV_COI))*100, 1)

##16S
unassigned_16S <- 
  filter(tax_16S_60, is.na(phylum)) %>% 
  pull(seq)

#proportion of sequences
round((length(unassigned_16S) / nrow(tax_16S_60))*100, 1)

#proportion of reads
round((sum(ASV_16S[,unassigned_16S]) / sum(ASV_16S))*100, 1)


  
```

We remove all sequences that have no taxonomic assignment at Phylum level with a (non-conservative) confidence value of at least 60%

COI:
This removes 39.3 % of the sequences accounting for 14.3% of all reads.

16S:
This removes 66.8 % of the sequences accounting for 9.5% of all reads.

1) check for sequences in the negative control

```{r}

## COI

# sequences in blank sample & assignment
blank_coi <- 
ASV_COI[filter(sampleInfo, ID == "CLEAN2" & Barcode == "COI") %>% pull(Sample),] %>% 
  pivot_longer(everything()) %>% 
  filter(value > 0) %>% 
  dplyr::rename(seq = name) %>% 
  left_join(tax_COI_60) %>% 
  arrange(desc(value))

blank_coi

#look at sequence prevalence 
select(ASV_COI, pull(blank_coi, seq)) %>% 
  tibble::rownames_to_column("Sample") %>% 
  pivot_longer(starts_with("seq")) %>% 
  mutate(value = value > 0) %>% 
  group_by(name) %>% 
  summarise(value = sum(value))


## 16S

# sequences in blank sample & assignment
blank_16S <- 
  ASV_16S[filter(sampleInfo, ID == "CLEAN2" & Barcode == "16S") %>% pull(Sample),] %>% 
  pivot_longer(everything()) %>% 
  filter(value > 0) %>% 
  dplyr::rename(seq = name) %>% 
  left_join(tax_16S_60) %>% 
  arrange(desc(value))

blank_16S


#prevalence
select(ASV_16S, pull(blank_16S, seq)) %>% 
  tibble::rownames_to_column("Sample") %>% 
  pivot_longer(starts_with("seq")) %>% 
  mutate(value = value > 0) %>% 
  group_by(name) %>% 
  summarise(value = sum(value))
```

COI: 
the blank sample contains 11 sequences for COI. Only 4 have a taxonomic assignments at Kingdom level. 2 have a certain species level assignment:  *Caradina Sp.* and *Corbicula fluminea*. We have confirmed *Caradina sp.* as lab contamination (100 % match, extraction on that species were performed during the same time in the lab as our samples). The sequences was present in all samples. *Corbicula fluminea* was only present in the blank but we cannot confirm the origin of this sequence. 

seq_0010, which is identified as a Coleoptera here, is in fact a human sequence. The Sequence is miss-assigned in the Sintax reference database (checked manually)

16S: 
the blank sample contains 9 sequences for 16S. Only one has a taxonomic assignment that is trustworthy, *Sus scrofa*, which is the wild boar / domestic pig and could be a contamination from the sterile sampling location as it only occurs in that sample. Seq_0009 is very likely the same Caradina contamination as it also occurs in (almost) all samples

All sequences present in the negative control are removed from all samples


2) check for Human sequences
--> would show up as primate in the 16S refDB as there are no human seqs in the reference database but also no relatives that are likely to show up in the samples
--> For 16S one sequence in the reference database, "Thrips flavidulus" is a human contaminant. Therefore many human sequences got assigned to this species.  We remove all instance of this Species. 


```{r}
humans_coi <- 
tax_COI_60 %>% 
  filter(order == "Primates") %>% 
  pull(seq)

humans_16S <- 
  tax_16S_60 %>% 
  filter(order == "Primates" | genus == "Thrips") %>% 
  pull(seq)
```

for COI we exclude 3 sequences

for 16S we exclude 24 sequences 

3) exclude potential contaminations

The following species are present in the data although they are very unlikely to have been present in the environment (do not occur in Sweden). However they have been handled in lab before and are likely to be contaminations. 


Clytie illunaris
Mycalesis anynana
Austrazenia pura 
Caradina sp. (already removed is it was present in the blank)
```{r}

contam_coi <- 
tax_COI_60 %>% 
  filter(species %in% c("Clytie illunaris",
                        "Mycalesis anynana",
                        "Austrazenia pura")) %>% 
  pull(seq)

#not present in 16S (or not identified to species level in 16S)
```

4) the most abundant Arthropod in both COI and 16S is a fly (Metopina pileata) and it's reads are concentrated in two samples (SI1 & SI3)

In those samples a small fly got caught in the tube (sucked in despite the net). The fly was removed but it dominated all the reads. We exclude it. 

```{r}
ASV_16S[, c("seq_0002", "seq_0645")] %>% 
  arrange(desc(seq_0002)) %>% 
  dplyr::slice(1:2) %>% 
  tibble::rownames_to_column("Sample") %>% 
  left_join(sampleInfo)

ASV_COI[, "seq_0001", drop = FALSE] %>% 
  arrange(desc(seq_0001)) %>% 
  dplyr::slice(1:2) %>% 
  tibble::rownames_to_column("Sample") %>% 
  left_join(sampleInfo)

fly_COI <- "seq_0001"
fly_16S <- c("seq_0002", "seq_0645")
```


#exclude sequences

```{r}
remove_COI <- unique(c(unassigned_coi, humans_coi, blank_coi$seq, contam_coi, fly_COI))
remove_16S <- unique(c(unassigned_16S, humans_16S, blank_16S$seq, fly_16S))

tax_COI_clean <- filter(tax_COI_60, !seq %in% remove_COI) 
tax_16S_clean <- filter(tax_16S_60, !seq %in% remove_16S) 

ASV_COI <- select(ASV_COI, -c(which(colnames(ASV_COI) %in% remove_COI))) 
ASV_16S <- select(ASV_16S, -c(which(colnames(ASV_16S) %in% remove_16S))) 
```



check for sequences prevalence in filtered COI data

```{r}
colSums(ASV_COI >0) %>% 
  sort(decreasing = TRUE) %>% 
  `[`(1:100) %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("seq") %>% 
  left_join(tax_COI_clean) %>% 
  mutate(rank = 1:n()) %>% 
  filter(`.` > 10) %>% 
  ggplot(aes(x = rank, y =`.`, label = species))+
  geom_point()+
  geom_label_repel()
  
filter(sampleInfo,  Sample %in%  rownames(ASV_COI)[which(ASV_COI$seq_0069	> 0)])
```

check for sequences prevalence in filtered 16S data

```{r}
colSums(ASV_16S >0) %>% 
  sort(decreasing = TRUE) %>% 
  `[`(1:100) %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("seq") %>% 
  left_join(tax_16S_clean) %>% 
  mutate(rank = 1:n()) %>% 
  ggplot(aes(x = rank, y =`.`, label = species))+
  geom_point()+
  geom_label_repel()
```


#phyla distribution

```{r}
Phylum_COI <- 
tax_COI_clean %>% 
  pivot_longer(-seq, names_to = "rank", values_to = "taxa") %>% 
  left_join(pivot_longer(prob_COI, -seq, names_to = "rank", values_to = "prob")) %>% 
  filter(rank == "phylum")


Phylum_COI %>% 
  mutate(taxa = ifelse(is.na(taxa), "not_assigned", taxa)) %>% 
  mutate(prob = round(prob, 1)) %>% 
  group_by(taxa) %>% 
  mutate(N = n()) %>% 
  group_by(taxa, prob) %>% 
  summarise(n = n(), N = unique(N)) %>%
  group_by(taxa) %>% 
  arrange(N, desc(prob)) %>% 
  mutate(taxa = factor(taxa, levels = unique(.$taxa))) %>% 
  mutate(alpha = case_when(prob >= 0.9 ~ 1,
                           prob < 0.9 & prob >= 0.8 ~ 0.8,
                           prob < 0.8 ~ 0.5)) %>% 
  ggplot(., aes(x = taxa, y = n, fill = prob, alpha = alpha))+
  geom_bar(position = "stack", stat = "identity")+
  coord_flip()+
  scale_fill_viridis_b()+
  scale_alpha(guide = 'none')+
  theme_minimal()
```

Arthropods

```{r}
Arthropds_COI <- 
tax_COI_clean %>% 
  filter(phylum == "Arthropoda") %>% 
  pivot_longer(-seq, names_to = "rank", values_to = "taxa") %>% 
  left_join(pivot_longer(prob_COI, -seq, names_to = "rank", values_to = "prob")) %>% 
  filter(rank == "class")

Arthropds_COI %>% 
  mutate(taxa = ifelse(is.na(taxa), "not_assigned", taxa)) %>% 
  mutate(prob = ifelse(taxa == "not_assigned", NA, prob)) %>% 
  mutate(prob = round(prob, 1)) %>% 
  group_by(taxa) %>% 
  mutate(N = n()) %>% 
  group_by(taxa, prob) %>% 
  summarise(n = n(), N = unique(N)) %>%
  group_by(taxa) %>% 
  arrange(N, desc(prob)) %>% 
  mutate(taxa = factor(taxa, levels = unique(.$taxa))) %>% 
  mutate(alpha = case_when(prob >= 0.9 ~ 1,
                           prob < 0.9 & prob >= 0.8 ~ 0.8,
                           prob < 0.8 ~ 0.5)) %>% 
  ggplot(., aes(x = taxa, y = n, fill = prob, alpha = alpha))+
  geom_bar(position = "stack", stat = "identity")+
  coord_flip()+
  scale_fill_viridis_b()+
  scale_alpha(guide = 'none')+
  theme_minimal()

```
Insects

```{r}
Insecta_COI <- 
tax_COI_clean %>% 
  filter(class == "Insecta") %>% 
  pivot_longer(-seq, names_to = "rank", values_to = "taxa") %>% 
  left_join(pivot_longer(prob_COI, -seq, names_to = "rank", values_to = "prob")) %>% 
  filter(rank == "order")
  

Insecta_COI %>% 
  mutate(taxa = ifelse(is.na(taxa), "not_assigned", taxa)) %>% 
  mutate(prob = ifelse(taxa == "not_assigned", NA, prob)) %>% 
  mutate(prob = round(prob, 1)) %>% 
  group_by(taxa) %>% 
  mutate(N = n()) %>% 
  group_by(taxa, prob) %>% 
  summarise(n = n(), N = unique(N)) %>%
  group_by(taxa) %>% 
  arrange(N, desc(prob)) %>% 
  mutate(taxa = factor(taxa, levels = unique(.$taxa))) %>%
  mutate(alpha = case_when(prob >= 0.9 ~ 1,
                           prob < 0.9 & prob >= 0.8 ~ 0.8,
                           prob < 0.8 ~ 0.5)) %>% 
  ggplot(., aes(x = taxa, y = n, fill = prob, alpha = alpha))+
  geom_bar(position = "stack", stat = "identity")+
  coord_flip()+
  scale_fill_viridis_b()+
  scale_alpha(guide = 'none')+
  theme_minimal()

```

16S

```{r}
Phylum_16S <- 
tax_16S_clean %>% 
  pivot_longer(-seq, names_to = "rank", values_to = "taxa") %>% 
  left_join(pivot_longer(prob_16S, -seq, names_to = "rank", values_to = "prob")) %>% 
  filter(rank == "phylum")


Phylum_16S %>% 
  mutate(taxa = ifelse(is.na(taxa), "not_assigned", taxa)) %>% 
  mutate(prob = ifelse(taxa == "not_assigned", NA, prob)) %>%  
  mutate(prob = round(prob, 1)) %>% 
  group_by(taxa) %>% 
  mutate(N = n()) %>% 
  group_by(taxa, prob) %>% 
  summarise(n = n(), N = unique(N)) %>%
  group_by(taxa) %>% 
  arrange(N, desc(prob)) %>% 
  mutate(taxa = factor(taxa, levels = unique(.$taxa))) %>% 
  mutate(alpha = case_when(prob >= 0.9 ~ 1,
                           prob < 0.9 & prob >= 0.8 ~ 0.8,
                           prob < 0.8 ~ 0.5)) %>% 
  ggplot(., aes(x = taxa, y = n, fill = prob, alpha = alpha))+
  geom_bar(position = "stack", stat = "identity")+
  coord_flip()+
  scale_fill_viridis_b()+
  scale_alpha(guide = 'none')+
  theme_minimal()
```

Arthropods

```{r}
Arthropds_16S <- 
tax_16S_clean %>% 
  filter(phylum == "Arthropoda") %>% 
  pivot_longer(-seq, names_to = "rank", values_to = "taxa") %>% 
  left_join(pivot_longer(prob_16S, -seq, names_to = "rank", values_to = "prob")) %>% 
  filter(rank == "class")

Arthropds_16S %>% 
 mutate(taxa = ifelse(is.na(taxa), "not_assigned", taxa)) %>% 
  mutate(prob = ifelse(taxa == "not_assigned", NA, prob)) %>% 
  mutate(prob = round(prob, 1)) %>% 
  group_by(taxa) %>% 
  mutate(N = n()) %>% 
  group_by(taxa, prob) %>% 
  summarise(n = n(), N = unique(N)) %>%
  group_by(taxa) %>% 
  arrange(N, desc(prob)) %>% 
  mutate(taxa = factor(taxa, levels = unique(.$taxa))) %>% 
  mutate(alpha = case_when(prob >= 0.9 ~ 1,
                           prob < 0.9 & prob >= 0.8 ~ 0.8,
                           prob < 0.8 ~ 0.5)) %>% 
  ggplot(., aes(x = taxa, y = n, fill = prob, alpha = alpha))+
  geom_bar(position = "stack", stat = "identity")+
  coord_flip()+
  scale_fill_viridis_b()+
  scale_alpha(guide = 'none')+
  theme_minimal()

```
Insects

```{r}
Insecta_16S <- 
tax_16S_clean %>% 
  filter(class == "Insecta") %>% 
  pivot_longer(-seq, names_to = "rank", values_to = "taxa") %>% 
  left_join(pivot_longer(prob_16S, -seq, names_to = "rank", values_to = "prob")) %>% 
  filter(rank == "order")
  

Insecta_16S %>% 
  mutate(taxa = ifelse(is.na(taxa), "not_assigned", taxa)) %>% 
  mutate(prob = ifelse(taxa == "not_assigned", NA, prob)) %>% 
  mutate(prob = round(prob, 1)) %>% 
  group_by(taxa) %>% 
  mutate(N = n()) %>% 
  group_by(taxa, prob) %>% 
  summarise(n = n(), N = unique(N)) %>%
  group_by(taxa) %>% 
  arrange(N, desc(prob)) %>% 
  mutate(taxa = factor(taxa, levels = unique(.$taxa))) %>%
  mutate(alpha = case_when(prob >= 0.9 ~ 1,
                           prob < 0.9 & prob >= 0.8 ~ 0.8,
                           prob < 0.8 ~ 0.5)) %>% 
  ggplot(., aes(x = taxa, y = n, fill = prob, alpha = alpha))+
  geom_bar(position = "stack", stat = "identity")+
  coord_flip()+
  scale_fill_viridis_b()+
  scale_alpha(guide = 'none')+
  theme_minimal()

```

## joined figure
```{r}

taxa_dist_16S <- 
Phylum_16S %>% 
  rbind(Arthropds_16S) %>% 
  rbind(Insecta_16S) %>% 
  mutate(taxa = ifelse(is.na(taxa), "not_assigned", taxa)) %>% 
  mutate(prob = ifelse(taxa == "not_assigned", NA, prob)) %>% 
  mutate(prob = round(prob, 1)) %>% 
  group_by(rank, taxa) %>% 
  mutate(N = n()) %>% 
  group_by(rank, taxa, prob) %>% 
  summarise(n = n(), N = unique(N)) %>% 
  mutate(marker = "16S") 

taxa_dist_COI <- 
Phylum_COI %>% 
  rbind(Arthropds_COI) %>% 
  rbind(Insecta_COI) %>% 
  mutate(taxa = ifelse(is.na(taxa), "not_assigned", taxa)) %>% 
  mutate(prob = ifelse(taxa == "not_assigned", NA, prob)) %>% 
  mutate(prob = round(prob, 1)) %>% 
  group_by(rank, taxa) %>% 
  mutate(N = n()) %>% 
  group_by(rank, taxa, prob) %>% 
  summarise(n = n(), N = unique(N)) %>% 
  mutate(marker = "COI") 

taxa_dist_joined <- 
rbind(taxa_dist_16S, taxa_dist_COI) %>% 
  group_by(rank, taxa) %>% 
  arrange(N, desc(prob)) %>% 
  mutate(taxa = factor(taxa, levels = unique(.$taxa))) %>% 
  mutate(alpha = case_when(prob >= 0.9 ~ 1,
                           prob < 0.9 & prob >= 0.8 ~ 0.8,
                           prob < 0.8 ~ 0.5))

phylum_plot <- 
  taxa_dist_joined %>% 
  filter(rank == "phylum") %>% 
  ggplot(., aes(x = taxa, y = n, fill = prob, alpha = alpha))+
  geom_bar(position = "stack", stat = "identity")+
  coord_flip()+
  facet_grid(rank ~ marker)+
  scale_fill_viridis_b()+
  scale_alpha(guide = 'none')+
  theme_minimal()+
  labs(x = "Phyla (all)\n ", y = "")+
  theme(
  strip.background = element_blank(),
  strip.text.y = element_blank()
)

class_plot <- 
  taxa_dist_joined %>% 
  filter(rank == "class") %>% 
  ggplot(., aes(x = taxa, y = n, fill = prob, alpha = alpha))+
  geom_bar(position = "stack", stat = "identity")+
  coord_flip()+
  facet_grid(rank ~ marker)+
  scale_fill_viridis_b()+
  scale_alpha(guide = 'none')+
  theme_minimal()+
  labs(x = "Classes from Phylum\nArthropoda", y = "")+
  theme(
  strip.background = element_blank(),
  strip.text.x = element_blank(),
  strip.text.y = element_blank()
)

order_plot <- 
  taxa_dist_joined %>% 
  filter(rank == "order") %>% 
  ggplot(., aes(x = taxa, y = n, fill = prob, alpha = alpha))+
  geom_bar(position = "stack", stat = "identity")+
  coord_flip()+
  facet_grid(rank ~ marker)+
  scale_fill_viridis_b()+
  scale_alpha(guide = 'none')+
  theme_minimal()+
  labs(x = "Order from Class\nInsecta", y = "number of species in each taxa")+
  theme(
  strip.background = element_blank(),
  strip.text.x = element_blank(),
  strip.text.y = element_blank()
)
  
ggarrange(
  phylum_plot, class_plot, order_plot, 
  common.legend = TRUE, legend = "right", ncol = 1, heights = c(14/14, 6/14, 13/14)
  )

ggsave(here("Figures", "Taxa_distribution.jpg"), width = 6, height = 8)
```


# comparision with field data

## moths survey

###rgbif

harmonize taxonomic names for moths
```{r}

Moths_spec <- unique(Moths$SciName)


gbif_specnames <- 
lapply(Moths_spec, function(x) {
  name_backbone(name = x, 
                rank = "species", 
                class = "Insecta")
}) %>% bind_rows()


Moths_taxa_full <-  
  gbif_specnames %>% 
  select(kingdom, phylum, class, order, family, genus, species) %>% 
  mutate(SciName = Moths_spec) %>% 
  relocate(SciName)
```

make species abundance table in OTU table format
```{r}
# names that differ in gbif
Moths_taxa_full %>% 
  filter(SciName != species) 

#replace
Moths <- 
Moths %>% 
  mutate(SciName = case_when(SciName == "Cirrhia icteritia" ~ "Xanthia icteritia",
                             TRUE ~ SciName))

Moths_light_ASV <- 
  Moths %>% 
  select(SciName, Antal, ID) %>% 
    mutate(rowID = 1:n()) %>% 
    group_by(SciName, ID) %>% 
    summarise(Antal = sum(Antal)) %>% # one case were a species was listed twice
    pivot_wider(names_from = SciName, values_from= Antal, values_fill = 0)
```

### moths from eDNA
```{r}
#COI
Moths_samples_COI <- sampleInfo[sampleInfo$Station == "ECO" & 
                              sampleInfo$Barcode == "COI" ,]$Sample

Moths_ASV_COI <- ASV_COI[Moths_samples_COI, ]
Moths_ASV_COI <- Moths_ASV_COI[,colSums(Moths_ASV_COI) > 0]

moths_seq_COI <- 
tax_COI_clean %>% 
  filter(seq %in% colnames(Moths_ASV_COI))

#16S
Moths_samples_16S<- sampleInfo[sampleInfo$Station == "ECO" & 
                              sampleInfo$Barcode == "16S" ,]$Sample

Moths_ASV_16S <- ASV_16S[Moths_samples_16S, ]
Moths_ASV_16S <- Moths_ASV_16S[,colSums(Moths_ASV_16S) > 0]

moths_seq_16S <- 
tax_16S_clean %>% 
  filter(seq %in% colnames(Moths_ASV_16S))
```

### euler moth 
```{r}
M_COI = moths_seq_COI %>% filter(order == "Lepidoptera") %>% 
  pull(species) %>% 
  na.omit() %>% 
  unique()


M_16S = moths_seq_16S %>% filter(order == "Lepidoptera") %>% 
  pull(species) %>% 
  na.omit() %>% 
  unique()

M_lt = Moths_taxa_full %>% filter(order == "Lepidoptera") %>% pull(SciName)

M_COI_M_16S  = dplyr::intersect(M_COI, M_16S)
M_COI_M_lt = dplyr::intersect(M_COI, M_lt)
M_16S_M_lt = dplyr::intersect(M_16S, M_lt)

v1 <- euler(c(A=length(M_COI), 
             B=length(M_16S), 
             C=length(M_lt),
             "A&B"=length(M_COI_M_16S),
             "A&C"=length(M_COI_M_lt),
             "B&C"=length(M_16S_M_lt)))

jpeg(here("Figures", "Euler_Moth_Lep.jpeg"))

plot(v1, 
     fills = alpha(c("#145DA0", "#7EA829", "#D11E00"),0.8),
     labels = c("COI", "16S", "light trapping"),
     quantities = TRUE,
     main = "Moth survey - only Moths" )

```

### euler Arthropods

```{r}

A_COI = moths_seq_COI %>% filter(phylum == "Arthropoda") %>% 
  pull(species) %>% 
  na.omit() %>% 
  unique()


A_16S = moths_seq_16S %>% filter(phylum == "Arthropoda") %>% 
  pull(species) %>% 
  na.omit() %>% 
  unique()

A_lt = Moths$SciName %>% unique()

A_COI_A_16S  = dplyr::intersect(A_COI, A_16S)
A_COI_A_lt = dplyr::intersect(A_COI, A_lt)
A_16S_A_lt = dplyr::intersect(A_16S, A_lt)

v2 <- euler(c(A=length(A_COI), 
             B=length(A_16S), 
             C=length(A_lt),
             "A&B"=length(A_COI_A_16S),
             "A&C"=length(A_COI_A_lt),
             "B&C"=length(A_16S_A_lt)))

jpeg(here("Figures", "Euler_Moth_Art.jpeg"))

plot(v2, 
     fills = alpha(c("#145DA0", "#7EA829", "#D11E00"),0.8),
     labels = c("COI", "16S", "light trapping"),
     quantities = TRUE,
     main = "Moth survey - all Arthropods" )

```


## Butterfly survey (F2F)

```{r}
#group over transects 
Insects_g <- 
  Insects %>% 
  group_by(IDnr, Habitat, Date, Species) %>% 
  summarise(Individuals = sum(Individuals))
```

###rgbif

harmonize taxonomic names for pollinators
```{r}

Pol_spec <- unique(Insects_g$Species)


gbif_specnames <- 
lapply(Pol_spec, function(x) {
  name_backbone(name = x, 
                rank = "species", 
                class = "Insecta")
}) %>% bind_rows()


Pol_taxa_full <-  
  gbif_specnames %>% 
  select(kingdom, phylum, class, order, family, genus, species) %>% 
  mutate(Species = Pol_spec) %>% 
  relocate(Species)
```


make species abundance table in OTU table format
```{r}
# names that differ in gbif
Pol_taxa_full %>% 
  filter(Species != species) %>% 
  select(Species, species)

#replace
Insects_g <- 
  Insects_g %>% 
  mutate(Species = case_when(
    Species == "Bombus lucorum coll." ~ "Bombus lucorum",
    Species == "Eumedonia eumedon" ~ "Aricia eumedon",
    TRUE ~ Species))

Pol_transect_ASV <- 
  Insects_g %>% 
  ungroup() %>% 
  select(IDnr, Species, Individuals) %>% 
    pivot_wider(names_from = Species, values_from= Individuals, values_fill = 0)
```


### Pollinators from eDNA
```{r}
#COI
Pol_samples_COI <- sampleInfo[sampleInfo$Station %in% c("SF", "SI", "SB") & 
                              sampleInfo$Barcode == "COI" ,]$Sample

Pol_ASV_COI <- ASV_COI[Pol_samples_COI, ]
Pol_ASV_COI <- Pol_ASV_COI[,colSums(Pol_ASV_COI) > 0]


Pol_seq_COI <- 
tax_COI_clean %>% 
  filter(seq %in% colnames(Pol_ASV_COI))

#16S
Pol_samples_16S <- sampleInfo[sampleInfo$Station %in% c("SF", "SI", "SB") & 
                              sampleInfo$Barcode == "16S" ,]$Sample

Pol_ASV_16S <- ASV_16S[Pol_samples_16S, ]
Pol_ASV_16S <- Pol_ASV_16S[,colSums(Pol_ASV_16S) > 0]

Pol_seq_16S <- 
tax_16S_clean %>% 
  filter(seq %in% colnames(Pol_ASV_16S))
```

### euler Pollinators 
```{r}
P_COI = Pol_seq_COI %>% filter(order == "Lepidoptera" | family == "Apidae") %>% 
  pull(species) %>% 
  na.omit() %>% 
  unique()


P_16S = Pol_seq_16S %>% filter(order == "Lepidoptera"| family == "Apidae") %>% 
  pull(species) %>% 
  na.omit() %>% 
  unique()

P_f2f = 
  Insects_g %>%
  filter(Individuals > 0) %>% 
  pull(Species) %>% 
  unique()

#exclude species not ID'd to species level
P_f2f <- P_f2f[!grepl("unidentified", P_f2f)]
P_f2f <-  P_f2f[-match("Bombus", P_f2f)]

P_COI_P_16S  = dplyr::intersect(P_COI, P_16S)
P_COI_P_f2f = dplyr::intersect(P_COI, P_f2f)
P_16S_P_f2f = dplyr::intersect(P_16S, P_f2f)

v3 <- euler(c(A=length(P_COI), 
             B=length(P_16S), 
             C=length(P_f2f),
             "A&B"=length(P_COI_P_16S),
             "A&C"=length(P_COI_P_f2f),
             "B&C"=length(P_16S_P_f2f)))

jpeg(here("Figures", "Euler_Pol_Pol.jpeg"))

plot(v3, 
     fills = alpha(c("#145DA0", "#7EA829", "#D11E00"),0.8),
     labels = c("COI", "16S", "transect counts"),
     quantities = TRUE,
     main = "Pollinator transects - Bees & Butterflies" )

```

### euler Arthropods

```{r}

A_COI = Pol_seq_COI %>% filter(phylum == "Arthropoda") %>% 
  pull(species) %>% 
  na.omit() %>% 
  unique()


A_16S = Pol_seq_16S %>% filter(phylum == "Arthropoda") %>% 
  pull(species) %>% 
  na.omit() %>% 
  unique()

A_COI_A_16S  = dplyr::intersect(A_COI, A_16S)
A_COI_P_f2f = dplyr::intersect(A_COI, P_f2f)
A_16S_P_f2f = dplyr::intersect(A_16S, P_f2f)

v4 <- euler(c(A=length(A_COI), 
             B=length(A_16S), 
             C=length(P_f2f),
             "A&B"=length(A_COI_A_16S),
             "A&C"=length(A_COI_P_f2f),
             "B&C"=length(A_16S_P_f2f)))

jpeg(here("Figures", "Euler_Pol_Art.jpeg"))

plot(v4, 
     fills = alpha(c("#145DA0", "#7EA829", "#D11E00"),0.8),
     labels = c("COI", "16S", "transects"),
     quantities = TRUE,
     main = "Pollinator transects - all Arthropods" )

dev.off()

```


#Bubblechart

## moths survey
```{r}

# Moths
#COI
Moths_pa_coi <- 
Moths_ASV_COI%>%
  tibble::rownames_to_column("ID") %>% 
  pivot_longer(starts_with("seq"),
               names_to = "seq", 
               values_to = "abundance") %>% 
  left_join(tax_COI_clean) %>% 
  group_by(kingdom, phylum,class, order,family,genus,species) %>% 
  summarise(abundance = sum(abundance>0)) %>% 
  filter(kingdom == "Animalia") %>% 
  filter(!is.na(species)) %>% 
  mutate(method = "COI")

#16S
Moths_pa_16S <- 
Moths_ASV_16S %>%
  tibble::rownames_to_column("ID") %>% 
  pivot_longer(starts_with("seq"),
               names_to = "seq", 
               values_to = "abundance") %>% 
  left_join(tax_16S_clean) %>% 
  group_by(kingdom, phylum,class, order,family,genus,species) %>% 
  summarise(abundance = sum(abundance >0)) %>% 
  filter(kingdom == "Animalia") %>%
  filter(!is.na(species)) %>%
  mutate(method = "16S")

# transects
Moths_pa_light<- 
  Moths_light_ASV %>% 
  pivot_longer(!one_of("ID"),
               names_to = "species", 
               values_to = "abundance") %>% 
  group_by(species) %>% 
  summarise(abundance = sum(abundance > 0)) %>% 
  filter(abundance > 0) %>% 
  left_join(Moths_taxa_full) %>% 
  select(phylum,class, order,family,genus,species, abundance) %>% 
  filter(!is.na(order)) %>% 
  mutate(method = "light trapping")

bind_rows(list(Moths_pa_16S, 
               Moths_pa_coi, 
               Moths_pa_light)) %>% 
  arrange(phylum, class, order, family, genus, species) %>% 
  mutate(species = factor(species, levels = unique(.$species))) %>% 
  mutate(order = case_when(is.na(order) ~ family,
                           TRUE ~ order)) %>% 
  mutate(order = factor(order, levels = unique(.$order))) %>% 
  ggplot(aes(x = species, y = method, colour = method, size = abundance))+
  geom_point()+
  facet_nested(phylum + order~., scales = "free", space = "free", nest_line = TRUE, switch = "y")+
  coord_flip()+
  scale_colour_manual(values = c("#145DA0", "#7EA829", "#D11E00"), guide = 'none')+
  scale_size_continuous(breaks = c(1,3,6,9,12,15))+
  guides(size = guide_legend(title.position="top", title.hjust = 0.5))+
  theme_bw() +
  theme(
    strip.text.y.left = element_text(angle = 0, size = 8),
    plot.margin = margin(0.5, 0.5, 0.5, 0.5, unit = "cm"),
   # plot.title = element_text(size = 15, face = "bold"),
    strip.placement = "outside",
    axis.title.y = element_blank(),
    axis.text = element_text(size = 8),
    legend.position = "bottom",
   strip.background = element_blank(),
        ggh4x.facet.nestline = element_line(colour = "blue")
   # panel.grid.major.y = element_blank(),
    )

ggsave(here("Figures", "Moths_bubble.pdf"), height = 14, width = 5)


```

## Butterfly survey
```{r}
# F2F

#COI
Pol_pa_coi <- 
Pol_ASV_COI %>%
  tibble::rownames_to_column("ID") %>% 
  pivot_longer(starts_with("seq"),
               names_to = "seq", 
               values_to = "abundance") %>% 
  left_join(tax_COI_clean) %>% 
  group_by(kingdom, phylum,class, order,family,genus,species) %>% 
  summarise(abundance = sum(abundance>0)) %>% 
  filter(kingdom == "Animalia") %>%
  filter(!is.na(species)) %>% 
  mutate(method = "COI")

#16S

Pol_pa_16S <- 
Pol_ASV_16S %>%
  tibble::rownames_to_column("ID") %>% 
  pivot_longer(starts_with("seq"),
               names_to = "seq", 
               values_to = "abundance") %>% 
  left_join(tax_16S_clean) %>% 
  group_by(kingdom, phylum,class, order,family,genus,species) %>% 
  summarise(abundance = sum(abundance >0)) %>% 
  filter(kingdom == "Animalia") %>%
  filter(!is.na(species)) %>%
  mutate(method = "16S")

# transects
Pol_pa_trans<- 
  Pol_transect_ASV %>% 
  pivot_longer(!one_of("IDnr"),
               names_to = "species", 
               values_to = "abundance") %>% 
  group_by(species) %>% 
  summarise(abundance = sum(abundance >0)) %>% 
  filter(abundance > 0) %>% 
  left_join(select(Pol_taxa_full, -Species)) %>% 
  select(phylum,class, order,family,genus,species, abundance) %>% 
  filter(!is.na(order)) %>% 
  mutate(method = "transect")

bind_rows(list(Pol_pa_16S, 
               Pol_pa_coi, 
               Pol_pa_trans)) %>% 
   arrange(phylum, class, order, family, genus, species) %>% 
  mutate(species = factor(species, levels = unique(.$species))) %>% 
  mutate(order = case_when(is.na(order) ~ family,
                           TRUE ~ order)) %>% 
  mutate(order = factor(order, levels = unique(.$order))) %>% 
  ggplot(aes(x = species, y = method, colour = method, size = abundance))+
  geom_point()+
  facet_nested(phylum + order~., scales = "free", space = "free", nest_line = TRUE, switch = "y")+
  coord_flip()+
  scale_colour_manual(values = c("#145DA0", "#7EA829", "#D11E00"), guide = 'none')+
  scale_size_continuous(breaks = c(1,3,6,9,12,15))+
  guides(size = guide_legend(title.position="top", title.hjust = 0.5))+
  theme_bw() +
  theme(
    strip.text.y.left = element_text(angle = 0, size = 8),
    plot.margin = margin(0.5, 0.5, 0.5, 0.5, unit = "cm"),
   # plot.title = element_text(size = 15, face = "bold"),
    strip.placement = "outside",
    axis.title.y = element_blank(),
    axis.text = element_text(size = 8),
    legend.position = "bottom",
   strip.background = element_blank(),
        ggh4x.facet.nestline = element_line(colour = "blue")
   # panel.grid.major.y = element_blank(),
    )

ggsave(here("Figures", "F2F_bubble.pdf"), height = 8, width = 5)


```

##IBA
```{r}
#COI
IBA_samples_COI <- sampleInfo[sampleInfo$Station == "IBA" & 
                              sampleInfo$Barcode == "COI" ,]$Sample

IBA_ASV_COI <- ASV_COI[IBA_samples_COI, ]
IBA_ASV_COI <- IBA_ASV_COI[,colSums(IBA_ASV_COI) > 0]

#16S
IBA_samples_16S <- sampleInfo[sampleInfo$Station == "IBA" & 
                              sampleInfo$Barcode == "16S" ,]$Sample

IBA_ASV_16S <- ASV_16S[IBA_samples_16S, ]
IBA_ASV_16S <- IBA_ASV_16S[,colSums(IBA_ASV_16S) > 0]


#COI
IBA_pa_coi <- 
IBA_ASV_COI %>%
  tibble::rownames_to_column("ID") %>% 
  pivot_longer(starts_with("seq"),
               names_to = "seq", 
               values_to = "abundance") %>% 
  left_join(tax_COI_clean) %>% 
  group_by(kingdom, phylum,class, order,family,genus,species) %>% 
  summarise(abundance = sum(abundance>0)) %>% 
  filter(kingdom == "Animalia") %>%
  filter(!is.na(species)) %>% 
  mutate(method = "COI")

#16S

IBA_pa_16S <- 
IBA_ASV_16S %>%
  tibble::rownames_to_column("ID") %>% 
  pivot_longer(starts_with("seq"),
               names_to = "seq", 
               values_to = "abundance") %>% 
  left_join(tax_16S_clean) %>% 
  group_by(kingdom, phylum,class, order,family,genus,species) %>% 
  summarise(abundance = sum(abundance >0)) %>% 
  filter(kingdom == "Animalia") %>%
  filter(!is.na(species)) %>%
  mutate(method = "16S")
```


##combined graph
```{r}
Bubble_data_long <- 
bind_rows(list(Pol_pa_16S, 
               Pol_pa_coi, 
               Pol_pa_trans)) %>% 
  mutate(study = "Pollinator site") %>% 
  bind_rows(., {bind_rows(
               Moths_pa_16S,
               Moths_pa_coi,
               Moths_pa_light) %>% 
       mutate(study = "Moths site")
      }) %>% 
  bind_rows(., {bind_rows(
               IBA_pa_coi,
               IBA_pa_16S) %>% 
       mutate(study = "Beech forest")
      }) %>% 
   arrange(phylum, class, order, family, genus, species) %>% 
  mutate(species = factor(species, levels = unique(.$species))) %>% 
  mutate(order = case_when(is.na(order) ~ family,
                           TRUE ~ order)) %>% 
  mutate(order = factor(order, levels = unique(.$order))) %>% 
  mutate(phylum = as.factor(phylum)) %>% 
  mutate(phylum = fct_relevel(phylum, "Annelida", after = Inf))

Bubble_data_LepHym <- 
  Bubble_data_long %>% 
  filter(order %in% c("Lepidoptera", "Hymenoptera"))

Bubble_data_other <- 
  Bubble_data_long %>% 
  filter(!order %in% c("Lepidoptera", "Hymenoptera"))
  
Other <- 
ggplot(Bubble_data_other, aes(x = species, y = method, fill = method, size = abundance))+
  geom_point(shape = 21, alpha = 0.8)+
  facet_nested(phylum + order~study, scales = "free", space = "free", nest_line = TRUE, switch = "y")+
  coord_flip()+
  labs(x = "", y = "", size = "prevalence")+
  scale_fill_manual(values = c("#145DA0", "#7EA829", "#D11E00", "#D11E00"), guide = 'none')+
  scale_size_continuous(breaks = c(1,3,6,9,12,15), range = c(1,5))+
  guides(size = guide_legend(nrow = 1))+
  theme_bw() +
  theme(
   strip.text.y.left = element_text(angle = 0, size = 8),
   strip.placement = "outside",
   axis.text = element_text(size = 8),
   axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
   axis.text.y = element_text(face = "italic"),
   legend.position = "bottom", 
   strip.background = element_blank(),
   ggh4x.facet.nestline = element_line(colour = "blue")
    )


LepHym <- 
ggplot(Bubble_data_LepHym, aes(x = species, y = method, fill = method, size = abundance))+
  geom_point(shape = 21, alpha = 0.8)+
  facet_nested(phylum + order~study, scales = "free", space = "free", nest_line = TRUE, switch = "y")+
  coord_flip()+
  labs(x = "", y = "", size = "prevalence")+
  scale_fill_manual(values = c("#145DA0", "#7EA829", "#D11E00", "#D11E00"), guide = 'none')+
  scale_size_continuous(breaks = c(1,3,6,9,12,15), range = c(1,5))+
  guides(size = guide_legend(nrow = 1))+
  theme_bw() +
  theme(
   strip.text.y.left = element_text(angle = 0, size = 8),
   strip.placement = "outside",
   axis.text = element_text(size = 8),
   axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
   axis.text.y = element_text(face = "italic"),
   legend.position = "bottom", 
   strip.background = element_blank(),
   ggh4x.facet.nestline = element_line(colour = "blue")
    )

ggsave(here("Figures", "Bubble_combined_Lep_Hym.pdf"), height = 35, width = 18, units = "cm", plot = LepHym)

ggsave(here("Figures", "Bubble_combined_other.pdf"), height = 35, width = 18, units = "cm", plot = Other)



```

#Diveristy 

## by site
_
```{r}

sampleInfo_stat <- 
  sampleInfo %>% 
  filter(!ID %in% c( "TEST", "CLEAN2")) %>% 
  mutate(Site = substr(Station, 1, 1)) %>% 
  split(.$Site)

ASV_16S_art <- ASV_16S[, colnames(ASV_16S) %in% 
                         tax_16S_clean[
                           tax_16S_clean$phylum ==
                             "Arthropoda",]$seq]

ASV_COI_art <- ASV_COI[, colnames(ASV_COI) %in% 
                         tax_COI_clean[
                           tax_COI_clean$phylum ==
                             "Arthropoda",]$seq]


all_16S <- 
lapply(sampleInfo_stat, function(x){
  
  specaccum(ASV_16S[rownames(ASV_16S) %in% x$Sample,])[c("sites", "richness", "sd")] %>% 
  as.data.frame()
  
}) %>% 
  bind_rows(.id = "Station") %>% 
  mutate(marker = "16S") %>% 
  mutate(group = "all")

art_16S <- 
lapply(sampleInfo_stat, function(x){
  
  specaccum(ASV_16S_art[rownames(ASV_16S_art) %in% x$Sample,])[c("sites", "richness", "sd")] %>% 
  as.data.frame()
  
}) %>% 
  bind_rows(.id = "Station") %>% 
  mutate(marker = "16S") %>% 
  mutate(group = "Arthropods")

all_COI <- 
lapply(sampleInfo_stat, function(x){
  
  specaccum(ASV_COI[rownames(ASV_COI) %in% x$Sample,])[c("sites", "richness", "sd")] %>% 
  as.data.frame()
  
}) %>% 
  bind_rows(.id = "Station") %>% 
  mutate(marker = "COI") %>% 
  mutate(group = "all")

art_COI <- 
lapply(sampleInfo_stat, function(x){
  
  specaccum(ASV_COI_art[rownames(ASV_COI_art) %in% x$Sample,])[c("sites", "richness", "sd")] %>% 
  as.data.frame()
  
}) %>% 
  bind_rows(.id = "Station") %>% 
  mutate(marker = "COI") %>% 
  mutate(group = "Arthropods")



rbind(all_16S, art_16S,all_COI,art_COI) %>% 
   mutate(Volume = case_when(
    Station == "E" ~ sites * 9, 
    Station == "S" ~ sites * 6, 
    Station == "I" ~ sites * 6)) %>% 
  mutate(Station = case_when(
    Station == "E" ~ "Moths site", 
    Station == "S" ~ "Pollinator site", 
    Station == "I" ~ "Beech forest")) %>% 
ggplot( aes(x = Volume, y= richness, fill = Station))+
  #geom_point()+
  geom_line(colour = "black")+
  geom_ribbon(aes(ymax = richness + sd, ymin = richness - sd), alpha = 0.5, colour = "NA")+
  facet_wrap(~group + marker, 
             scales = "free",
             labeller =
               label_wrap_gen(multi_line=FALSE))+
  theme_minimal()+
  scale_fill_brewer(palette = "Set1")+
  theme(panel.grid.major = element_blank())+
  labs(y = "sequence richness", x = bquote('cumulative volume of air'~(m^3)),
       title = "species accumulation curves")

ggsave(here("Figures", "Specaccum.pdf"), width = 6, height = 4)

ggsave(here("Figures", "Specaccum.jpeg"), width = 6, height = 4)
  
```

## by habitat
```{r}
sampleInfo <- 
  sampleInfo %>% 
  left_join(site_codes) %>% 
  left_join(F2F_meta)

sampleInfo_stat <- 
  sampleInfo %>% 
  filter(grepl("^S", ID), perl = TRUE) %>% 
  split(.$habitat)

ASV_16S_art <- ASV_16S[, colnames(ASV_16S) %in% 
                         tax_16S_clean[
                           tax_16S_clean$phylum ==
                             "Arthropoda",]$seq]

ASV_COI_art <- ASV_COI[, colnames(ASV_COI) %in% 
                         tax_COI_clean[
                           tax_COI_clean$phylum ==
                             "Arthropoda",]$seq]


all_16S <- 
lapply(sampleInfo_stat, function(x){
  
  specaccum(ASV_16S[rownames(ASV_16S) %in% x$Sample,])[c("sites", "richness", "sd")] %>% 
  as.data.frame()
  
}) %>% 
  bind_rows(.id = "Station") %>% 
  mutate(marker = "16S") %>% 
  mutate(group = "all")

art_16S <- 
lapply(sampleInfo_stat, function(x){
  
  specaccum(ASV_16S_art[rownames(ASV_16S_art) %in% x$Sample,])[c("sites", "richness", "sd")] %>% 
  as.data.frame()
  
}) %>% 
  bind_rows(.id = "Station") %>% 
  mutate(marker = "16S") %>% 
  mutate(group = "Arthropods")

all_COI <- 
lapply(sampleInfo_stat, function(x){
  
  specaccum(ASV_COI[rownames(ASV_COI) %in% x$Sample,])[c("sites", "richness", "sd")] %>% 
  as.data.frame()
  
}) %>% 
  bind_rows(.id = "Station") %>% 
  mutate(marker = "COI") %>% 
  mutate(group = "all")

art_COI <- 
lapply(sampleInfo_stat, function(x){
  
  specaccum(ASV_COI_art[rownames(ASV_COI_art) %in% x$Sample,])[c("sites", "richness", "sd")] %>% 
  as.data.frame()
  
}) %>% 
  bind_rows(.id = "Station") %>% 
  mutate(marker = "COI") %>% 
  mutate(group = "Arthropods")



rbind(all_16S, art_16S,all_COI,art_COI) %>% 
  mutate(Volume = sites *2) %>% 
ggplot( aes(x = Volume, y= richness, fill = Station))+
  #geom_point()+
  geom_line(colour = "black")+
  geom_ribbon(aes(ymax = richness + sd, ymin = richness - sd), alpha = 0.5, colour = "NA")+
  facet_wrap(~marker+group, 
             scales = "free",
             labeller =
               label_wrap_gen(multi_line=FALSE))+
  theme_minimal()+
  scale_fill_brewer(palette = "Set1")+
  theme(panel.grid.major = element_blank())+
  labs(y = "sequence richness", x = bquote('cumulative volume of air'~(m^3)),
       title = "species accumulation curves")
  
```

#NMDS

##make phyloseq objects
```{r}

#COI
tax_COI_clean_m <- as.matrix(tax_COI_clean[,-1])
rownames(tax_COI_clean_m) <- tax_COI_clean$seq

sampleInfo <- 
  as.data.frame(sampleInfo) %>% 
  filter(!Station %in% c("TEST", "CLEAN"))
  
rownames(sampleInfo) <- sampleInfo$Sample

sampleInfo$Station %>% unique

ps_coi <- phyloseq(otu_table(ASV_COI+1, taxa_are_rows = FALSE),
                   tax_table(tax_COI_clean_m),
                   sample_data(sampleInfo))

#16S
tax_16S_clean_m <- as.matrix(tax_16S_clean[,-1])
rownames(tax_16S_clean_m) <- tax_16S_clean$seq

ps_16S <- phyloseq(otu_table(ASV_16S+1, taxa_are_rows = FALSE),
                   tax_table(tax_16S_clean_m),
                   sample_data(sampleInfo))
```

## all seqs
### COI

```{r}
d2_coi <- phyloseq_to_deseq2(ps_coi, ~ Station)

VST_coi <- t(assay(varianceStabilizingTransformation(d2_coi, blind = T)))

VST_coi[VST_coi < 0 ]  <- 0

dist_coi <- vegdist(VST_coi, distance = "bray")

NMDS_coi <- metaMDS(dist_coi, autotransform = FALSE, trace = 0)

points_coi <- NMDS_coi$points %>% 
  data.frame() %>% 
  tibble::rownames_to_column(var = "Sample") %>% 
  left_join(.,sampleInfo) 

hull <- points_coi %>%
  group_by(Station) %>%
  dplyr::slice(chull(MDS1, MDS2))

ggplot(points_coi, aes(x = MDS1, y =MDS2, colour = Station, 
                   fill = Station))+
  geom_point(size = 2)+
  #geom_text(aes(label = sample), nudge_x= 0.02, size = 2)+
  theme_bw()+
  ggtitle("Species composition - bray curtis\nCOI all sequences")+
  scale_colour_brewer(palette = "Set1", name = "Station")+
  scale_fill_brewer(palette = "Set1", name = "Station")+
  scale_shape_manual(values = c(16,24))+
  geom_polygon(data = hull, alpha = 0.3)+
  scale_alpha_manual(values = c(0.4, 0))

ggsave(here("Figures", "NMDS_COI.jpg"), width = 4, height = 3)

```

### stats
```{r}
#overall station effect
adonis2(dist_coi ~ Station, data = filter(sampleInfo, Sample %in% rownames(VST_coi)))

#station effect for Pollinator site
VST_PS <- VST_coi[rownames(VST_coi) %in% sampleInfo[grepl("S", sampleInfo$Station), ]$Sample,]

VST_PS <- VST_PS[, colSums(VST_PS) > 0]

sampleInfo <- 
sampleInfo %>% left_join(site_codes) %>% left_join(F2F_meta)

adonis2(VST_PS ~ Station + habitat, data = filter(sampleInfo, Sample %in% rownames(VST_PS)), dist = "bray")


```


### 16S

```{r}
d2_16S <- phyloseq_to_deseq2(ps_16S, ~ Station)

VST_16S <- t(assay(varianceStabilizingTransformation(d2_16S, blind = T)))

VST_16S[VST_16S < 0 ]  <- 0

dist_16S <- vegdist(VST_16S, distance = "bray")

NMDS_16S <- metaMDS(dist_16S, autotransform = FALSE, trace = 0)

points_16S <- NMDS_16S$points %>% 
  data.frame() %>% 
  tibble::rownames_to_column(var = "Sample") %>% 
  left_join(.,sampleInfo) 

hull <- points_16S %>%
  group_by(Station) %>%
  dplyr::slice(chull(MDS1, MDS2))

#p1 <- 
ggplot(points_16S, aes(x = MDS1, y =MDS2, colour = Station, 
                   fill = Station))+
  geom_point(size = 2)+
  #geom_text(aes(label = sample), nudge_x= 0.02, size = 2)+
  theme_bw()+
#  ggtitle("16S composition - bray curtis 2017")+
  scale_colour_brewer(palette = "Set1", name = "Treatment")+
  scale_fill_brewer(palette = "Set1", name = "Treatment")+
  scale_shape_manual(values = c(16,24))+
  geom_polygon(data = hull, alpha = 0.3)+
  scale_alpha_manual(values = c(0.4, 0))

#ggsave(filename = "../figures/NMDS_16S_2017.pdf",plot = p1, width = 8, height = 6)
```


## numbers for results

```{r}
#number of seqs
tax_COI_clean %>% nrow()
tax_16S_clean %>% nrow()

#number of Fungi
tax_COI_clean %>% filter(phylum == "Basidiomycota") %>% nrow()
tax_COI_clean %>% filter(phylum == "Ascomycota") %>% nrow()

#number of phyla
tax_COI_clean %>% pull(phylum) %>% unique %>% length
tax_16S_clean %>% pull(phylum) %>% unique


#number of Arthropods
tax_COI_clean %>% filter(phylum == "Arthropoda") %>% nrow()
tax_16S_clean %>% filter(phylum == "Arthropoda") %>% nrow()


#number of insects
tax_16S_clean %>% filter(class == "Insecta") %>% nrow()
tax_COI_clean %>% filter(class == "Insecta") %>% nrow()

#number of unique insects id'd to species level
Insect_uniq_coi <- 
tax_COI_clean %>% 
  filter(class == "Insecta") %>% 
  filter(!is.na(species)) %>% 
  pull(species) %>% 
  unique() 

length(Insect_uniq_coi)

Insect_uniq_16S <- 
tax_16S_clean %>% 
  filter(class == "Insecta") %>%
  filter(!is.na(species)) %>%
  pull(species) %>% 
  unique() 

length(Insect_uniq_16S)

unique(c(Insect_uniq_coi, Insect_uniq_16S)) %>% length()

#number of Arthropods at species level
COI_art_spec <- 
tax_COI_clean %>% 
   filter(phylum == "Arthropoda") %>% 
    filter(!is.na(species)) %>% 
    pull(species) %>% 
    unique()

length(COI_art_spec)


art_spec_16S <- 
tax_16S_clean %>% 
   filter(phylum == "Arthropoda") %>% 
    filter(!is.na(species)) %>% 
    pull(species) %>% 
    unique()

length(art_spec_16S)

unique(c(COI_art_spec, art_spec_16S)) %>% length()

#percent fungal reads
ASV_fung <- tax_COI_clean %>% filter(phylum %in% c("Basidiomycota", "Ascomycota")) %>% pull(seq)

sum(ASV_COI[,ASV_fung])/sum(ASV_COI)                             
                                     
```

