---
title: "EMBL_16S"
author: "Fabian Roger"
date: "6/18/2020"
output: html_document
---

```{r}
library(dplyr)
library(tidyr)
library(DECIPHER)
library(future.apply)
library(readr)
library(here)
library(ggplot2)
library(RCurl) 
library(httr)
```


#EMBL
##download EMBL data

make directory to store files downloaded from EMBL (~120GB)
```{bash, eval= FALSE}
#cd /Volumes/Elements/
#mkdir EMBL
```


list all nt files from EMBL
```{r}
url <- "ftp://ftp.ebi.ac.uk/pub/databases/embl/release/std/"


results <- getURL(url,verbose=TRUE,ftp.use.epsv=TRUE, dirlistonly = TRUE)
results <- unlist(strsplit(results, "\n"))

results <- results[grepl("_std_", results)]

lapply(results, function(x) strsplit(x, "_")[[1]][3]) %>% 
  unlist() %>% 
  table()

```

we download all but the `Synthetic` (syn), `Transgenic` (tgn), `Human` (hum) and `Mus musculus` (mus) and `Environmental samples` (env) and `Unclassified` (unc)

While we will need some human & mice sequences in the reference db we can add them later and don't need to download *all* sequences from mice and humans for these two species. The Environmental samples don't have taxonomic information and thus don't help us to assign a species

we download the following files:
```{r}

results <- results[grepl("_fun_|_inv_|_mam_|_phg_|_pln_|_pro_|_rod_|_vrl_|_vrt_", results)]

results
```


```{r}
diskurl <- "/Volumes/Elements/EMBL/"

for (i in results) {
  
  if( ! file.exists(paste(diskurl, i, sep = ""))){
    
    message(paste("downloading file ", which(results == i), "out of", length(results), " :", i))
    
    GET(
      paste(url, i, sep = ""),
      write_disk(paste(diskurl, i, sep = "")),
      progress())
  }
}


```

##download taxdump
the taxdump gives the relationships between the different phylogenetic levels (child and parents)
```{r}
taxurl <- "ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz"

GET(taxurl,
    write_disk("/Volumes/Elements/taxdump.tar.gz"),
    progress())


```

#ObiTools3

We use ObiTools3. For download instructions see here: https://git.metabarcoding.org/obitools/obitools3/wikis/home

##import into DMS
importing the files in the DMS (copy paste into bash window)

A DMS is how OBITools needs the files, sort of a database
```{bash}

cd /Applications/OBITools/obitools3
source obi3-env/bin/activate
cd /Volumes/Elements/
obi import --embl /Volumes/Elements/EMBL /Volumes/Elements/EMBL_dms/EMBL_ref

```

import taxdump into DMS
```{bash}
obi import --taxdump /Volumes/Elements/taxdump.tar.gz /Volumes/Elements/EMBL_dms/taxonomy/my_tax
```

#16S

Marquina, Daniel, Anders F. Andersson, and Fredrik Ronquist. “New Mitochondrial Primers for Metabarcoding of Insects, Designed and Evaluated Using in Silico Methods.” Molecular Ecology Resources 19, no. 1 (January 2019): 90–104. https://doi.org/10.1111/1755-0998.12942.

```{r}
#16S
data.frame(`.` = c( "sequence","direction","length"),
           Chiar16SF = c( "TARTYCAACATCGRGGTC", "foward", nchar("TARTYCAACATCGRGGTC")),
           Chiar16SR = c( "CYGTRCDAAGGTAGCATA", "reverse", nchar("CYGTRCDAAGGTAGCATA"))) %>% 
  kable(caption = "primer sequences")

```

##activate obitools
```{bash}
cd /Applications/OBITools/obitools3
source obi3-env/bin/activate
```


##ecopcr

ecopcr goes through all seqeunces in the EMBL_ref database which contains all sequences downloaded from EMBL. 

It the extracts all seq that match the forward and the reverse primer, with 2 missmatches allowed. Note that it doensn't just grep fro the primers but simulates and *in-silico* PCR and returns the sequences that would be amplified. 

I filter the resulting sequences to only keep those between 300 and 450 bps

```{bash}
obi ecopcr -e 2 -l 300  -L 450 -F TARTYCAACATCGRGGTC -R CYGTRCDAAGGTAGCATA --taxonomy airDNA/taxonomy/my_tax airDNA/embl_refs airDNA/Chiar16S_refs
```

number of found sequences

```{bash}
cd /Applications/OBITools/obitools3
source obi3-env/bin/activate
obi count /Volumes/Elements/EMBL_dms/Chiar16S_refs
```

No we filter out sequences to keep only those that have full taxonomic annotation, i.e. a family, genus and species level assignment. The other sequences are discarded. 

```{bash}
obi grep --require-rank=species --require-rank=genus --require-rank=family --taxonomy /Volumes/Elements/EMBL_dms/taxonomy/my_tax /Volumes/Elements/EMBL_dms/Chiar16S_refs /Volumes/Elements/EMBL_dms/Chiar16S_refs_clean
```

number of sequences left

```{bash}
obi count /Volumes/Elements/EMBL_dms/Chiar16S_refs_clean
```

2020-04-28 16:41:53,665 [count : INFO ]  obi count
246293 246293

make sure sequence are annotated at all taxonomic levels

##annotate refDB

```{bash}
cd /Applications/OBITools/obitools3
source obi3-env/bin/activate

obi annotate --with-taxon-at-rank superkingdom --with-taxon-at-rank kingdom --with-taxon-at-rank phylum --with-taxon-at-rank class --with-taxon-at-rank order --with-taxon-at-rank order --with-taxon-at-rank family --taxonomy /Volumes/Elements/EMBL_dms/taxonomy/my_tax /Volumes/Elements/EMBL_dms/Chiar16S_refs_clean /Volumes/Elements/EMBL_dms/Chiar16S_refs_clean_annotated
```


##export 

export sequence information for each sequence
```{bash}

cd /Applications/OBITools/obitools3
source obi3-env/bin/activate


obi export --tab-output --header /Volumes/Elements/EMBL_dms/Chiar16S_refs_clean_annotated > ~/Documents/01_Work/01_Research/15_eDNA/Data/16S_refDB.txt

```

export sequences as fasta file
```{bash}
cd /Applications/OBITools/obitools3
source obi3-env/bin/activate

obi export --fasta /Volumes/Elements/EMBL_dms/Chiar16S_refs_clean_annotated > /Volumes/Elements/BLAST/16S_refDB.fasta
```


## rename seqs

We have exported the sequence metadata (including taxonomic annotation) into the `16S_refDB.txt` file. We therefore can keep only the identifier as sequence names and rename after clustering with consensus taxonomy. 

```{r}
refseq16S <- readDNAStringSet("/Volumes/Elements/BLAST/16S_refDB.fasta")

names(refseq16S) <- gsub("([A-Z,0-9]+).+", "\\1", names(refseq16S))

writeXStringSet(refseq16S, "/Volumes/Elements/BLAST/16S_refDB.fasta")
```


##cluster sequences

```{bash}
/Applications/vsearch/bin/vsearch --threads 4 --cluster_fast /Volumes/Elements/BLAST/16S_refDB.fasta --centroids /Volumes/Elements/BLAST/16S_refDB_99.fasta --id 0.99 --uc /Volumes/Elements/BLAST/16S_refDB_99.txt
```


90483535 nt in 246293 seqs, min 300, max 450, avg 367
Clusters: 80532 Size min 1, max 1785, avg 3.1
Singletons: 48175, 19.6% of seqs, 59.8% of clusters

## find consensus taxonomy

We now have 3 files: 
 (1) the fasta file with the centroid sequences from the 99% clustering
 (2) the text file telling us what sequences were clustered to what centroid
 (3) the metadata file giving us the taxonomic annotation (from EMBL) of all sequences (the centroids and the ones that have been clustered)
 
 We read in all 3 files 

```{r}

# import centroid seqeunces
refseq16S_99 <- readDNAStringSet("/Volumes/Elements/BLAST/16S_refDB_99.fasta")

# import cluster information 
Cluster_ID_16S <-  read_tsv("/Volumes/Elements/BLAST/16S_refDB_99.txt", col_names = FALSE)

#filter only sequences that have been clusterd
Cluster_ID_16S <- 
Cluster_ID_16S %>% filter(X1 != "C") %>% 
  dplyr::rename(seq = X9, Type = X1, match = X10, pident = X4) %>% 
  select(seq, Type, match, pident) %>% 
  filter(Type == "H" ) %>% 
  arrange(match)

#import taxonomic annotation from embl
refDB_16S <- read_tsv(here("Data", "16S_refDB.txt"))

refDB_16S <- refDB_16S %>% 
  select(ID, ends_with("_name"))

colnames(refDB_16S) <- gsub("(.?)_name","\\1", colnames(refDB_16S))

refDB_16S <- select(refDB_16S, ID, superkingdom, kingdom, phylum, class, order, family, genus, species)

```

Here I generate a named list of all centroid sequences, which contains the phylogenetic annotation for the centroid sequence itself as well as all sequences clustered to that centroid

```{r}
Cluster_list <- 
split(Cluster_ID_16S, Cluster_ID_16S$match)

plan(multisession)

Cluster_list_dfs <-
future_lapply(names(Cluster_list), function(x){
  
  clust_seq <- Cluster_list[[x]]$seq
  refDB_16S[refDB_16S$ID %in% c(x, clust_seq),]
  
})

names(Cluster_list_dfs) <- names(Cluster_list)

Cluster_list_dfs[1]


```

which clusters have fully coherent taxonomic assignments (Superkingdom to species)?

```{r}

plan(multisession)

cluster_ntax <- 
future_lapply(names(Cluster_list_dfs), function(x){
  
  cluster <- x
  
  x <- Cluster_list_dfs[[x]]
  
  nspec <- x %>% 
    select(-ID) %>% 
    distinct() %>% 
    nrow()
  
  ngen <- x %>% 
    select(-ID, - species) %>% 
    distinct() %>% 
    nrow()
  
  nfam <- x %>% 
    select(-ID, - species, -genus) %>% 
    distinct() %>% 
    nrow()
  
  tibble(cluster = cluster,
         ngen = ngen,
         nspec = nspec,
         nfam = nfam)
  
}) %>% bind_rows()

```

There are 32357 clusters (non-singletons) were we can assess if the taxonomy is coherent. 

Of those, 253 have incongruities at the family level, 1513 have incongruities at the genus level and 9781 have incongruities at the species level. 

incongruities can arise in different ways: 

(1) 16S is a very conserved marker. It is likely that clustering at 99% will create clusters that genuinely have sequences from different species and even genera (see e.g. "AB034824")

(2) The are a lot of variation for species names were some sequences are only identified as sp. of cf. or have a number code / specimen id. This is likely the biggest culprit for species incongruities

(3) misslabeled / missidentified sequences. These are hard(er) to detect if the missidentification is at species or genus level but if sequences disagree at family level it is clear that some sequences in the cluster are misslabeled. 


We define the consensus taxonomy as follows:

(1) if Family name differs, we exclude the cluster
(2) if Genus name differs, we replace the Genus and the Family name with NA
(3) if species name differs, we remove entries that contain 'sp.' in species name. If it still differs, we use genus sp. as species name

```{r}
#singleton cluster
cl_singl <- names(refseq16S_99)[!names(refseq16S_99) %in% Cluster_ID_16S$match]

# cluster with congruent taxonomy
cl_sp_c <- cluster_ntax %>% filter(nspec == 1) %>% pull(cluster)

# cluster were family name differs
cl_fam <- cluster_ntax %>% filter(nfam > 1) %>% pull(cluster)

#cluster were genus name differ
cl_gen <- cluster_ntax %>% filter(nfam == 1 & ngen > 1) %>% pull(cluster)

#cluster were only species name differ
cl_spec <- cluster_ntax %>% filter(ngen == 1 & nspec > 1) %>% pull(cluster)
```

```{r}
# taxonomy for singleton clusters
cl_singl <- refDB_16S %>% filter(ID %in% cl_singl)

# taxonomy for congruent clusters
cl_sp_c <- refDB_16S %>% filter(ID %in% cl_sp_c)

# taxonomy for incongruent genera, with genus and sp set to NA
cl_gen <- 
  refDB_16S %>% 
  filter(ID %in% cl_gen) %>% 
  mutate(genus = NA_character_, species = NA_character_)

# taxonomy with incongruent species
plan(multisession)

cl_sp_df <- 
future_lapply(cl_spec, function(x){
  
  df <- Cluster_list_dfs[[x]]
  
  tax <- 
  df %>% 
    filter(!grepl("sp.", species))
  
  ntax <- tax %>% pull(species) %>% unique %>% length()
  
  if(ntax == 1){
    df <- filter(df, ID == x)
  } else{
    df <- filter(df, ID == x) %>% 
      mutate(species = paste(genus, "sp.", sep = " "))
  }
}) %>% bind_rows()

Cluster_tax <- 
  bind_rows(cl_singl, cl_sp_c, cl_gen, cl_sp_df)
```

now we have a consensus taxonomy for all cluster centroids. Now we 

(1) filter out clusters without consensus taxonomy (family disagreement)
(2) rename sequences with consensus taxonomy
(3) export sequences

```{r}
#exclude clusters
refseq16S_99_cons <- 
  refseq16S_99[names(refseq16S_99) %in% Cluster_tax$ID]

#make taxonomic annotation
Cluster_tax <- 
  Cluster_tax %>% 
  mutate(species = gsub(" ", "_", species)) %>% 
  mutate(taxonomy = paste(ID, superkingdom, kingdom, phylum, class, order, family, genus, species, sep = ";"))

#rename sequences with taxonomc annotation
names(refseq16S_99_cons) <- 
  Cluster_tax$taxonomy[ match( names(refseq16S_99_cons), Cluster_tax$ID ) ]

#export seqs
writeXStringSet(refseq16S_99_cons, "/Volumes/Elements/BLAST/16S_refDB_99_wtax.fasta")

```



